firedrake/variational_solver.py

```diff
@@ -242,7 +242,7 @@ def set_transfer_manager(self, manager):
 
     @PETSc.Log.EventDecorator()
     @NonlinearVariationalSolverMixin._ad_annotate_solve
-    def solve(self, bounds=None):
+    def solve(self, bounds=None, rhs=None):
```

Comment:
Is this the right API, or should the RHS be attached to the `NonlinearVariationalProblem`? It seems odd to add symbolic information about the problem at this late stage.

Reply:
Additionally, this must surely require a corresponding adjoint change.

Reply:
I agree about the adjoint change.

My use case: I'm implementing FAS algorithms for variational inequalities, and I'm setting up firedrake `NonlinearVariationalProblem`s and `NonlinearVariationalSolver`s to do the smoothing on each level and on the coarse grid. FAS constructs a 'correction' to the nonlinear problem on each grid, to ensure that the solution of the corrected coarse grid problem is the injection of the fine grid solution. These are the right-hand sides I want to supply here, and are as far as I know the only common use case for passing something nonzero to `SNESSolve`. Since it changes each iteration, it would be less convenient to make it a property of the `NonlinearVariationalProblem`.